fractal_tasks_core/cellpose_segmentation.py

```diff
@@ -55,7 +58,223 @@
+    :param ROI_table_obs: ``obs`` attribute of the AnnData table for the
+                          primary-label ROIs; this is used (together with
+                          ``ROI_index``) to extract ``label_value``.
+    :param ROI_index: index of the current ROI, which is used to extract
+                      ``label_value`` from ``ROI_table_obs``.
+    """
+
+    # If use_masks=False, return immediately
+    if not use_masks:
+        return (image_array, None, None)
+
+    logger.info(f"[preprocess_cellpose_input] {image_array.shape=}")
+    logger.info(f"[preprocess_cellpose_input] {region=}")
+
+    # Check that image data are 4D (CZYX) - FIXME issue 340
+    if not image_array.ndim == 4:
+        raise ValueError(
+            "preprocess_cellpose_input requires a 4D "
+            f"image_array argument, but {image_array.shape=}"
+        )
+
+    # Check that all optional arguments are non-None
+    optional_args = (
+        region,
+        current_label_path,
+        ROI_index,
+        ROI_table_path,
+    )
+
+    if None in optional_args:
+        raise ValueError(
+            f"Wrong arguments for preprocess_cellpose_input:\n"
+            f"{use_masks=}\n"
+            f"{current_label_path=}\n, "
+            f"{ROI_index=}\n"
+            f"{region=}\n"
+            f"{ROI_table_path=}"
+        )
+
+    # Load the ROI table and its metadata attributes
+    ROI_table = ad.read_zarr(ROI_table_path)
+    ROI_table_obs = ROI_table.obs
+    attrs = zarr.group(ROI_table_path).attrs
+    if not attrs["type"] == "ngff:region_table":
+        raise ValueError("Wrong attributes for {ROI_table_path}:\n{attrs}")
+    label_relative_path = attrs["region"]["path"]
+    column_name = attrs["instance_key"]
+    print(label_relative_path)
+    print(column_name)
+
+    # Check that ROI_table_obs has the right column and extract label_value
+    if column_name not in ROI_table_obs.columns:
+        raise ValueError(
+            'In preprocess_cellpose_input, "{column_name}" '
+            f" missing in {ROI_table_obs.columns=}"
+        )
+    label_value = int(ROI_table_obs[column_name][ROI_index])
+
+    # Load primary/current label arrays (lazily), and check shapes
+    primary_label_path = str(
+        Path(ROI_table_path).parent / label_relative_path / "0"
+    )
+    logger.critical(f"{primary_label_path=}")
+    primary_label_array = da.from_zarr(primary_label_path)
+    current_label_array = da.from_zarr(current_label_path)
+    logger.info(
+        f"[preprocess_cellpose_input] {primary_label_path=}, "
+        f"{primary_label_array.shape=}"
+    )
+    logger.info(
+        f"[preprocess_cellpose_input] {current_label_path=}, "
+        f"{current_label_array.shape=}"
+    )
+
+    # Load ROI data for current label array
+    current_label_region = current_label_array[region].compute()
+
+    # Load ROI data for primary label array, with or without upscaling
+    if primary_label_array.shape != current_label_array.shape:
```

Comment:
Do we check here whether the primary label array is in higher resolution? That would be an issue and out-of-scope.

Reply:
Yes, taken care of in `convert_region_to_low_res`.